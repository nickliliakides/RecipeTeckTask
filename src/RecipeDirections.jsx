import React from 'react';
import Loader from './Loader';

const RecipeDirections = ({ recipe, loading }) => {
  return recipe && !loading ? (
    <div className='recipe'>
      <figure className='recipe__fig'>
        <img src={recipe.source_url} alt='Tomato' className='recipe__img' />
        <h1 className='recipe__title'>
          <span>{recipe.title}</span>
        </h1>
      </figure>
      <div className='recipe__details'>
        <div className='recipe__info'>
          <svg className='recipe__info-icon'>
            <use href='img/icons.svg#icon-stopwatch'></use>
          </svg>
          <span className='recipe__info-data recipe__info-data--minutes'>
            {`${recipe.prep_time_min} + ${recipe.cook_time_min}`}
          </span>
          <span className='recipe__info-text'> minutes (prep + cook)</span>
        </div>
      </div>

      <div className='recipe__directions'>
        <h2 className='heading-2'>How to cook it</h2>
        {recipe.directions.map((dir, i) => (
          <p key={i} className='recipe__directions-text'>
            {dir}
          </p>
        ))}
        <hr />
        <p className='recipe__directions-last-text'>
          This recipe was carefully designed and tested by
          <span className='recipe__by'>&nbsp;{recipe.author.name}</span>. Please
          check out more recipes at their website.
        </p>
        <a
          className='btn-small recipe__btn'
          href={recipe.author.url}
          target='_blank'
          rel='noreferrer'
        >
          <span>Directions</span>
          <svg className='search__icon'>
            <use href='img/icons.svg#icon-triangle-right'></use>
          </svg>
        </a>
      </div>
    </div>
  ) : (
    <Loader />
  );
};

export default RecipeDirections;
