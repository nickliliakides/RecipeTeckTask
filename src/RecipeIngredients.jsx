import React, { useEffect, useState } from 'react';
import RecipeIngredientItem from './RecipeIngredientItem';
const Fraction = require('fractional').Fraction;

const RecipeIngredients = ({ recipe, loading }) => {
  const [ingredients, setIngredients] = useState([]);
  const [currentServings, setCurrentServings] = useState();

  useEffect(() => {
    if(recipe) {
      if(!currentServings) setCurrentServings(recipe.servings)
      /* Here is handled all the logic to slplit the ingredient in two strings(quantity and the ingredient itself). If there is a fraction need to be checked as well, and myltiplied with the factor(default servings / current servings) to do the calculation everytime the user increases or decreases the servings quantity  */
      const ingredientArray = recipe.ingredients.map((ing) => {
        const ingredientsArray = ing.split(' ');
        let count = ingredientsArray[0];
        if (ingredientsArray[1].match(/^\d/)) {
          count = `${count} ${ingredientsArray[1]}`;
          ingredientsArray.splice(0, 2);
        } else {
          ingredientsArray.splice(0, 1);
        }
        if (currentServings !== recipe.servings) {
          if (count.includes('/')) {
            if (count.includes(' ')) {
              const newCountArray = count.split(' ');
              const fractionArray = newCountArray[1].split('/');
              const countNum = new Fraction(newCountArray[0]).add(
                new Fraction(fractionArray[0], fractionArray[1])
              );
  
              count = countNum
                .multiply(new Fraction(currentServings, recipe.servings))
                .toString();
            } else {
              const fractionArray = count.split('/');
              count = new Fraction(fractionArray[0], fractionArray[1])
                .multiply(new Fraction(currentServings, recipe.servings))
                .toString();
            }
          } else {
            count = new Fraction(count)
              .multiply(new Fraction(currentServings, recipe.servings))
              .toString();
          }
        }
        const ingredient = ingredientsArray.join(' ');
  
        return { count, ingredient };
      });

      setIngredients(ingredientArray);
    }
   
  }, [recipe, currentServings]);

  return recipe && !loading ? (
    <div className='shopping'>
      <div className='ingredients'>
        <h2 className='heading-2'>Ingredients</h2>
        <div className='recipe__info'>
          <svg className='recipe__info-icon'>
            <use href='img/icons.svg#icon-man'></use>
          </svg>
          <span className='recipe__info-data recipe__info-data--people'>
            {currentServings}
          </span>
          <span className='recipe__info-text'> servings</span>
          <div className='recipe__info-buttons'>
            <button
              disabled={currentServings === 1}
              className='btn-tiny btn-decrease'
              onClick={() =>
                setCurrentServings((currentServ) => {
                  return currentServ - 1;
                })
              }
            >
              <svg>
                <use href='img/icons.svg#icon-circle-with-minus'></use>
              </svg>
            </button>
            <button
              className='btn-tiny btn-increase'
              onClick={() =>{
                if(currentServings < 20) {
                  setCurrentServings((currentServ) => {
                    return currentServ + 1;
                  })
                } else {
                  alert('20 people are the greater number to calculate the ingredients')
                }
              }

              }
            >
              <svg>
                <use href='img/icons.svg#icon-circle-with-plus'></use>
              </svg>
            </button>
          </div>
        </div>

        <ul className='recipe__ingredient-list'>
          {ingredients.map((ing, i) => (
            <RecipeIngredientItem key={i} item={ing} />
          ))}
        </ul>
      </div>

      <div className='categories'>
        <h2 className='heading-2 cat'>Categories</h2>
        <ul className='recipe__tags__list'>
          {recipe.tags.map((tag, i) => (
            <li key={i}>
              <p>{tag}</p>
            </li>
          ))}
        </ul>
      </div>
    </div>
  ) : <div className='loading'>Loading ...</div>
};

export default RecipeIngredients;
