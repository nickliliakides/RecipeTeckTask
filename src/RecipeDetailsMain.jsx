import React, { Fragment } from 'react';
import RecipeDirections from './RecipeDirections';
import RecipeIngredients from './RecipeIngredients';

const RecipeDetailsMain = ({ recipe, loading }) => {
  return (
    <Fragment>
      <RecipeDirections recipe={recipe} loading={loading} />
      <RecipeIngredients recipe={recipe} loading={loading} />
    </Fragment>
  );
};

export default RecipeDetailsMain;
