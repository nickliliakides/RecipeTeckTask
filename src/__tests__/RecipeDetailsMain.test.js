import React from 'react';
import { shallow } from 'enzyme';
import RecipeDetailsMain from '../RecipeDetailsMain';
import RecipeDirections from '../RecipeDirections';
import RecipeIngredients from '../RecipeIngredients';

let wrapper;

const defaultProps = {
  loading: false,
  setLoading: jest.fn(),
};

describe('<RecipeDetailsMain /> component', () => {
  beforeEach(() => {
    wrapper = shallow(<RecipeDetailsMain {...defaultProps} />);
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it('should matches its snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render the RecipeDirections component', () => {
    expect(wrapper.find(RecipeDirections).length).toEqual(1);
  });

  it('should render the RecipeIngredients component', () => {
    expect(wrapper.find(RecipeIngredients).length).toEqual(1);
  });
});
