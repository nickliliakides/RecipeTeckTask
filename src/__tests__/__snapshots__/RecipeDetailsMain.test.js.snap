// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RecipeDetailsMain /> component should matches its snapshot 1`] = `
<Fragment>
  <RecipeDirections
    loading={false}
  />
  <RecipeIngredients
    loading={false}
  />
</Fragment>
`;
